simpl/middleware/errors.py

```diff
@@ -0,0 +1,144 @@
+            "traceback": "Traceback (most recent call last): ... ",
+        }
+
+
+    An "expected" error if the server is in debug mode:
+
+        # raise SimplHTTPError(body='Hey!', status=405)
+        {
+            "code": 405,
+            "description": "Hey!",
+            "exception": "SimplHTTPError('Hey!',)",
+            "message": "Method Not Allowed",
+            "traceback": "Traceback (most recent call last): ... ",
+        }
+
+
+    Unexpected error without debug mode:
+
+        # raise ValueError()
+        {
+            "code": 500,
+            "description": "We're sorry, something went wrong.",
+            "message": "Internal Server Error"
+        }
+
+
+    An "expected" error without debug mode:
+
+        # raise SimplHTTPError(body='Hey!', status=405)
+        {
+            "code": 405,
+            "description": "Hey!",
+            "message": "Method Not Allowed"
+        }
+"""
+
+import logging
+import traceback as tb_mod
+
+import bottle
+
+from simpl import exceptions as simpl_exc
+from simpl import rest
+
+LOG = logging.getLogger(__name__)
+
+
+class FormatExceptionMiddleware(object):
+
+    """Format outgoing exceptions.
+
+    Uses and is compatible-with bottle exception formatting.
+
+    - Handle Bottle Exceptions (even when catchall=False).
+    - Handle SimplHTTPError
+    - Fail-safe to a generic error (unexpected_error)
+
+    The code in this middleware is meant to be generic.
+    Don't catch and translate fancy exceptions here, do
+    it in the application logic or a separate middleware and
+    raise a bottle.HTTPError or a SimplHTTPError.
+    """
+
+    def __init__(self, app, conf=None):
+        """Initialize class."""
+        self.app = app
+        self.config = conf
+        LOG.debug("Added middleware: %s", type(self).__name__)
+
+    def __call__(self, environ, start_response):
+        """Catch exceptions and format them based on config."""
+        try:
+            LOG.warning("Ohai from middleware.")
+            response = self.app(environ, start_response)
+            LOG.warning("NO EXCEPTION.")
+            return response
+        except bottle.HTTPError as error:
+            LOG.error("Formatting a bottle exception.",
+                      exc_info=error)
+
+            rest.format_error_response(error)
+            start_response(error.status_line, error.headerlist)
+            return error
+        except simpl_exc.SimplHTTPError as error:
+            LOG.error("Formatting a SimplHTTPError exception.",
+                      exc_info=error)
+            error = bottle.HTTPError(
+                status=error.status_code, body=error.body,
+                exception=error.exception or error,
+                traceback=error.traceback or tb_mod.format_exc())
+            rest.format_error_response(error)
+            start_response(error.status_line, error.headerlist)
+            return error
+        except Exception as error:  # pylint: disable=W0703
```

Comment:
How will this impact SystemExit or KeyboardInterrupts?

What are the chances of an Exception being raised that is not inheriting from Exception? That may be bad practice, but should we also catch it using, maybe, an `except:`?